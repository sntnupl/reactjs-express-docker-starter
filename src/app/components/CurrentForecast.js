import React from 'react';
import PropTypes from 'prop-types';
import {ForecastCity} from '../services/WeatherService';

const ForecastSummary = ({data, error}) => {
    if (error && error.msg) {
        return (
            <p>{error.msg}</p>
        );
    }

    if (!data) return null;
    const forecastItems = data.list.map((item) => {
        return (
            <div className="forecast-card" key={item.dt.toString()}>
                <ul>
                    <li>{item.dt_txt}</li>
                    <li><i className="fa fa-check" aria-hidden="true"/> {item.weather[0].description}</li>
                    <li><i className="fa fa-thermometer-empty" aria-hidden="true"/> {item.main.temp} deg celcius</li>
                </ul>
            </div>
        );
    });
    return (
        <div className="forecast-card-container">{forecastItems}</div>
    );
};


ForecastSummary.propTypes = {
    error: PropTypes.shape({
        msg: PropTypes.string
    }),
    data: PropTypes.shape({
        name: PropTypes.string,
        country: PropTypes.string,
        list: PropTypes.arrayOf(PropTypes.object),
    }),
};

/*
export const getForecastFirstData = (loc) => {
    return ForecastCity(loc).then((resp) => {
        return {
            cityName: resp.data.city.name,
            data: {
                temp: resp.data.list[0].main.temp,
                min: resp.data.list[0].main.temp_min,
                max: resp.data.list[0].main.temp_max,
            }
        };
    }).catch((err) => {
        return err.msg;
    });
};
*/


class CurrentForecast extends React.Component {
    static Summary = ForecastSummary;

    constructor(props) {
        super(props);
    }

    refCity = null;
    state = {
        error: {
            msg: undefined
        },
        isFetchingData: false,
        forecastData: undefined,
    };

    handleForecastData = (forecast) => {
        //console.log(`Forecast for ${location.trim()}: ${JSON.stringify(resp)}`);
        if (this.refCity) this.refCity.value = '';
        const {name, country} = forecast.data.city;
        this.setState(() => (
            {
                isFetchingData: false,
                forecastData: {
                    name,
                    country,
                    list: forecast.data.list,
                }
            }
        ));
    };

    handleForecastError = (err) => {
        console.warn(`error: ${JSON.stringify(err)}`);
        this.setState(() => (
            {
                isFetchingData: false,
                error: {
                    msg: 'failed to fetch forecast'
                }
            }
        ));
    };

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.refCity.value) return;

        this.setState(() => {
            return {
                error: {msg: undefined},
                isFetchingData: true,
                forecastData: undefined,
            };
        });

        const location = this.refCity.value;
        ForecastCity(location.trim())
            .then(resp => this.handleForecastData(resp))
            .catch(err => this.handleForecastError(err));
    };

    render = () => {
        const {isFetchingData, error, forecastData} = this.state;

        return (
            <div className="forecast">
                <div className={forecastData ? "forecast-form-secondary" : "forecast-form-primary"}>
                    <form onSubmit={this.onSubmit}>
                        <input
                            type="text"
                            ref={(el) => {this.refCity = el}}
                            placeholder="city name (, country name)"/>
                        {!isFetchingData
                            ? <button type="submit">Submit</button>
                            : <button type='submit' disabled>Fetching..</button>
                        }
                    </form>
                </div>
                {forecastData &&
                    <div className="forecast-location">
                        <p>Weather Forecast: <b>{forecastData.name}, {forecastData.country}</b></p>
                    </div>
                }
                <ForecastSummary error={error} data={forecastData} />
            </div>
        );
    };

}


export default CurrentForecast;