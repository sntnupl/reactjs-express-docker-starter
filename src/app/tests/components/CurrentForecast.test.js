import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
jest.mock('../../services/WeatherService');
import CurrentForecast from '../../components/CurrentForecast';
//import CurrentForecast, {getForecastFirstData} from '../../components/CurrentForecast';


const dataForecast = {
    data: {
        city: {
            name: 'Kolkata',
            country: 'IN'
        },
        list: [
            {
                dt_txt: '2017-12-27 09:00',
                dt: 14008327,
                main: {
                    temp: 14,
                    temp_min: 12,
                    temp_max: 25,
                },
                weather: [{description: 'haze'}],
            }, {
                dt_txt: '2017-12-27 12:00',
                dt: 14008327,
                main: {
                    temp: 18,
                    temp_min: 12,
                    temp_max: 25,
                },
                weather: [{description: 'mist'}],
            }, {
                dt_txt: '2017-12-27 15:00',
                dt: 14008327,
                main: {
                    temp: 21,
                    temp_min: 12,
                    temp_max: 25,
                },
                weather: [{description: 'sunny'}],
            },
        ],
    }
};

test('should render CurrentForecast correctly', () => {
    const wrapper = shallow(<CurrentForecast/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should disable submit button once forecast starts getting fetched', () => {
    const wrapper = shallow(<CurrentForecast/>);
    expect(wrapper.find('button [disabled]').length).toBe(0);
    wrapper.setState({ isFetchingData: true });
    expect(wrapper.find('button [disabled]').length).toBe(1);
});

test('should handle forecast data', () => {
    const wrapper = shallow(<CurrentForecast/>);
    wrapper.setState({
        isFetchingData: true,
        forecastData: undefined,
    });

    wrapper.instance().handleForecastData(dataForecast);

    expect(wrapper.state().isFetchingData).toEqual(false);
    expect(wrapper.state().forecastData).toBeDefined();
});

test('should handle forecast error', () => {
    const wrapper = shallow(<CurrentForecast/>);
    wrapper.setState({
        isFetchingData: true,
        forecastData: undefined,
        error: {
            msg: undefined
        },
    });

    wrapper.instance().handleForecastError({});

    expect(wrapper.state().isFetchingData).toEqual(false);
    expect(wrapper.state().forecastData).toBeUndefined();
    expect(wrapper.state().error.msg.length).toBeGreaterThan(0);
});


test('should not render forecasts on error', () => {
    const wrapper = shallow(<CurrentForecast.Summary/>);
    wrapper.setProps({ error: { msg: 'test error' } });
    expect(wrapper.find('.weather-forecasts').length).toBe(0);
});

test('should render all forecast data', () => {
    const wrapper = shallow(<CurrentForecast.Summary />);
    wrapper.setProps({
        data: {
            name: 'test city',
            country: 'test country',
            list: dataForecast.data.list,
        },
    });
    expect(wrapper.find('.forecast-card').length).toBe(dataForecast.data.list.length);
});

/*
test('promises must be kept', async () => {
    expect.assertions(1);
    const resp = await getForecastFirstData('Kolkata');
    expect(resp.cityName).toEqual('Kolkata');
});*/
