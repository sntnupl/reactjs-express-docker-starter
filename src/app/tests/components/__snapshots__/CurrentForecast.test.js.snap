// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CurrentForecast correctly 1`] = `
<div
  className="forecast"
>
  <div
    className="forecast-form-primary"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        placeholder="city name (, country name)"
        type="text"
      />
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
  <ForecastSummary
    error={
      Object {
        "msg": undefined,
      }
    }
  />
</div>
`;
